tsl/src/remote/txn_resolve.c

```diff
@@ -18,14 +19,26 @@
 RemoteTxnResolution
 remote_txn_resolution(Oid foreign_server, const RemoteTxnId *transaction_id)
 {
-	if (remote_txn_is_still_in_progress(transaction_id->xid))
-		/* transaction still ongoing; don't know it's state */
+	if (remote_txn_is_still_in_progress_on_access_node(transaction_id->xid))
+		/* transaction still ongoing; don't know its state */
 		return REMOTE_TXN_RESOLUTION_UNKNOWN;
 
+	/*
+	 * If an entry exists in the "remote_txn" table and is visible then it means
+	 * that the transaction committed on the AN
+	 */
 	if (remote_txn_persistent_record_exists(transaction_id))
-		return REMOTE_TXN_RESOLUTION_COMMT;
+		return REMOTE_TXN_RESOLUTION_COMMIT;
 
-	return REMOTE_TXN_RESOLUTION_ABORT;
+	/*
+	 * We need to check the Postgres machinery to ensure that this txn has indeed
+	 * aborted before declaring so. If it's not aborted and not committed as per
+	 * above then it has to be a concurrent ongoing transaction.
+	 */
+	if (TransactionIdDidAbort(transaction_id->xid))
```

Comment:
I think this only works if the transaction is still on record in the clog. I am wondering if there is a situation where a data node has been down for a long time and vacuum has removed the record of the XID on the access node, thus this might return `REMOTE_TXN_RESOLUTION_UNKNOWN` and the record will never be cleaned up because it is believed to still be ongoing?


Therefore, I think it is a bit strange to check for an explicit abort when we implement "presumed abort". Given that we implement "presumed abort" it should be the default unless we have an explicit record of it either being in process or committed.

Reply:
Went back to the earlier logic of declaring abort.